swupd/delta.go

```diff
@@ -0,0 +1,190 @@
+				if err != nil {
+					failedChan <- f
+				}
+			}
+		}(statedir)
+	}
+
+	for _, file := range deltas {
+		select {
+		case deltaQueue <- file:
+		case f := <-failedChan:
+			failed = append(failed, f)
+		}
+	}
+
+	close(deltaQueue)
+	wg.Wait()
+
+	if len(failedChan) > 0 {
+		f := <-failedChan
+		failed = append(failed, f)
+	}
+
+	return failed, nil
+}
+
+func consolidateDeltaFiles(manifest *Manifest, from uint32, c config) ([]*File, error) {
+	var files []*File
+
+	if manifest == nil {
+		return nil, nil
+	}
+
+	for _, file := range manifest.Files {
+		if file.Version <= from {
+			continue
+		}
+		if file.DeltaPeer == nil {
+			continue
+		}
+		if file.Type != TypeFile ||
+			file.DeltaPeer.Type != TypeFile {
+			continue
+		}
+
+		deltadir := filepath.Join(c.outputDir, fmt.Sprintf("%d", file.Version), "delta")
+		fromToString := fmt.Sprintf("%d-%d-%s-%s", file.DeltaPeer.Version, file.Version, file.DeltaPeer.Hash, file.Hash)
+
+		deltaFile := filepath.Join(deltadir, fromToString)
+
+		// Only add to list if the delta file does not already exist on the system
+		if _, err := os.Stat(deltaFile); os.IsNotExist(err) {
+			// Only create deltas for files bigger than 200 bytes
+			fullname := filepath.Join(c.imageBase, fmt.Sprintf("%d", file.Version), "full"+file.Name)
+			file.Info, _ = os.Stat(fullname)
+			if file.Info.Size() > 200 {
+				files = append(files, file)
+			}
+		}
+
+	}
+
+	return files, nil
+}
+
+func createDelta(file *File, statedir string, fromVersion uint32, fromHash Hashval) error {
+	var c config
+	c, err := getConfig(statedir)
+	if err != nil {
+		return err
+	}
+
+	// File without peers cannot have deltas
+	if file.DeltaPeer == nil {
+		return nil
+	}
+
+	// We only support deltas between two regular files for now
+	// TODO: Support directory -> file in the future, this is a complex case
+	if file.Type != TypeFile || file.DeltaPeer.Type != TypeFile {
+		return nil
+	}
+
+	// By definition we cannot make deltas for files that are deleted
+	if file.Status == StatusDeleted {
+		return nil
+	}
+
+	// To files cannot be ghosted
+	if file.Status == StatusGhosted {
+		return nil
+	}
+
+	if sameFile(file, file.DeltaPeer) {
+		return nil
+	}
+
+	newfile := filepath.Join(c.imageBase, fmt.Sprintf("%d", file.Version), "full"+file.Name)
+	original := filepath.Join(c.imageBase, fmt.Sprintf("%d", fromVersion), "full"+file.Name)
+
+	deltadir := filepath.Join(c.outputDir, fmt.Sprintf("%d", file.Version), "delta")
+	fromToString := fmt.Sprintf("%d-%d-%s-%s", fromVersion, file.Version, fromHash, file.Hash)
+
+	// Files to create and validate deltas with
+	deltafile := filepath.Join(deltadir, fromToString)
+	testnewfile := filepath.Join(deltadir, "."+fromToString+".testnewfile")
+
+	// Shell out to bsdiff for now...false = don't print to screen
+	if err := helpers.RunCommand(false, "bsdiff", original, newfile, deltafile); err != nil {
+		_ = os.Remove(deltafile) // Might have retuned FULLDL
```

Comment:
Typo "retuned". Also comment is unclear, what is FULLDL?

Reply:
The Clear Linux BSDiff returns `1` as a hint to swupd-server that the delta it created is larger than the new file created after applying that delta. We call that hint FULLDL ("full download"), and swupd-server will make the fullfile (instead of the delta) available for download instead by adding it to the packs.

For reference, see https://github.com/clearlinux/swupd-server/blob/3cf2ade205cd4ef3189929655b72173752360f83/src/delta.c#L80

Reply:
Thanks for the explanation. I suggest a short version of it to be in a comment in this code.